src/vaults/vaults.service.ts

```diff
@@ -115,48 +115,73 @@ export class VaultsService {
+    const requestedVault = await chain.vaults.getVault(address);
+    const vaultBlobId = getVaultEntityId(chain, requestedVault);
     const dataKey = toChartDataKey(HistoricVaultSnapshotModel.NAMESPACE, vaultBlobId, timeframe);
-
     return queryVaultCharts(dataKey);
   }
 
+  /**
+   *
+   * @param vault
+   * @param chain
+   * @param timestamps
+   * @returns
+   */
   async getVaultChartDataByTimestamps(
-    vaultAddr: string,
-    network: Network,
+    vault: string,
+    chain: Chain,
     timestamps: number[],
   ): Promise<HistoricVaultSnapshotModel[]> {
-    // from smaller to greater, we need this order for search to work correctly
-    const sortedTimestamps = timestamps.sort((a, b) => (a > b ? 1 : -1));
-
-    const vaultBlobId = HistoricVaultSnapshotModel.formBlobId(vaultAddr, network);
-    const dataKey = toChartDataKey(HistoricVaultSnapshotModel.NAMESPACE, vaultBlobId, ChartTimeFrame.Max);
+    // validate vault request is correct and valid
+    const requestedVault = await chain.vaults.getVault(vault);
 
-    const vaultChartData = await queryVaultCharts(dataKey);
+    // sort timestamps in ascending order for searching
+    const timestampSort = (a: number, b: number) => (a > b ? 1 : -1);
+    let sortedTimestamps = timestamps.sort(timestampSort);
 
+    // TODO: we should consider making this a map and allowing ourselves to do validation against requested timestamps
+    // construct relevant persistence and search criteria
     const snapshots: Set<HistoricVaultSnapshotModel> = new Set();
+    const vaultBlobId = getVaultEntityId(chain, requestedVault);
 
-    if (vaultChartData.length === 0) {
-      return Array.from(snapshots);
-    }
+    // iterate over all valid charting granularities to ensure discovery of all timestamps
+    for (const timeframe of CHART_GRANULARITY_TIMEFRAMES) {
+      const dataKey = toChartDataKey(HistoricVaultSnapshotModel.NAMESPACE, vaultBlobId, timeframe);
 
-    // from smaller to greater, we need this order for search to work correctly
-    const preparedVaultsChartData = vaultChartData.reverse();
+      const vaultChartData = await queryVaultCharts(dataKey);
 
-    for (const timestamp of sortedTimestamps) {
-      const snapshotItem = preparedVaultsChartData.find((i) => Number(i.timestamp) >= timestamp);
+      if (vaultChartData.length === 0) {
+        continue;
+      }
 
-      if (snapshotItem) {
-        if (!snapshotItem.pricePerFullShare && snapshotItem.ratio) {
-          snapshotItem.pricePerFullShare = snapshotItem.ratio;
+      // from smaller to greater, we need this order for search to work correctly
+      const preparedVaultsChartData = vaultChartData.reverse();
+      const remainingTimestamps = [];
+
+      // iterate over timestamps updating the map as snapshots are found
+      for (const timestamp of sortedTimestamps) {
+        const snapshotItem = preparedVaultsChartData.find((i) => Number(i.timestamp) >= timestamp);
```

Comment:
does this always find just 1 tiemstamp?

Reply:
yes that is what find does, undefined or one

Reply:
i see comment now on required presort of that list